server/contexts/ctxerr/ctxerr.go

```diff
@@ -13,87 +13,190 @@ package ctxerr
 
 import (
 	"context"
+	"encoding/json"
 	"errors"
 	"fmt"
 	"strings"
 	"time"
-
-	"github.com/fleetdm/fleet/v4/server/errorstore"
-	"github.com/rotisserie/eris" //nolint:depguard
+	//nolint:depguard
 )
 
 type key int
 
 const errHandlerKey key = 0
 
-// Cause returns the root error in err's chain.
-func Cause(err error) error {
-	// Until we use only ctxerr.Wrap, we have a mix of pkg/errors, fmt.Errorf
-	// and eris.Wrap (via ctxerr.Wrap). pkg/errors.Cause looks for a Cause()
-	// method, while eris.Cause looks for the stdlib-compliant Unwrap(). So
-	// implement a custom Cause that checks for both until a root is found.
+// Defining here for testing purposes
+var nowFn = time.Now
+
+// FleetError is the error implementation used by this package.
+type FleetError struct {
+	msg   string      // error message to be prepended to cause
+	stack StackTracer // stack trace where this error was created
+	cause error       // original error that caused this error if non-nil
+	data  map[string]interface{}
+}
 
-	var cerr interface {
-		Cause() error
+type fleetErrorJSON struct {
+	Message string                 `json:",omitempty"`
+	Data    map[string]interface{} `json:",omitempty"`
+	Stack   []string               `json:",omitempty"`
+}
+
+// Error implements the error interface.
+func (e *FleetError) Error() string {
+	if e.cause == nil {
+		return e.msg
 	}
+	return fmt.Sprintf("%s: %s", e.msg, e.cause.Error())
+}
+
+// Unwrap implements the error Unwrap interface introduced in go1.13.
+func (e *FleetError) Unwrap() error {
+	return e.cause
+}
+
+// MarshalJSON implements the marshaller interface, giving us control on how
+// errors are json-encoded
+func (e *FleetError) MarshalJSON() ([]byte, error) {
+	return json.Marshal(fleetErrorJSON{
+		Message: e.msg,
+		Data:    e.data,
+		Stack:   e.stack.List(),
+	})
+}
+
+// New creates a new error with the given message.
+func New(ctx context.Context, msg string) *FleetError {
+	stack := NewStack(1)
+	err := &FleetError{msg, stack, nil, nil}
+	ensureCommonMetadata(ctx, err)
+	return err
+}
+
+// NewWithData creates a new error with the given message and attaches
+// aditional data to it.
+func NewWithData(ctx context.Context, msg string, data map[string]interface{}) *FleetError {
+	stack := NewStack(1)
+	err := &FleetError{msg, stack, nil, data}
```

Comment:
ah that's a really good catch. @mna since this is a `map[string]interface{}` we should have to do a recursive copy. Do you have any thoughts on doing this vs changing the signature to `map[string]string` for the sake of simplicity?

a `map[string]string` will push some complexity to the consumer tough, but should we start there and see how people use it? (if we use it at all?)

Reply:
I think the metadata is used only when marshalling to json for storage, right? If so we could marshal it right away and store it as `json.RawMessage`, that would address any reference issue.